## week 2-3 preprocessing homework draft

digma, ronachel marie of h3101

In [105]:
import pandas as pd

# load datasets

customers = pd.read_csv('../../data/raw/customer_demographics_contaminated.csv')
transactions = pd.read_csv('../../data/raw/customer_transactions_contaminated.csv')
interactions = pd.read_csv('../../data/raw/social_media_interactions_contaminated.csv')

### customer demographics dataset preprocessing

In [106]:
# inspect dataset
print("First 20 rows: ")
display(customers.head(20))

print("\nDataset Information: ")
display(customers.info())

print("\nDataset Dimensions: ")
display(customers.shape)

print("\nDataset Columns: ")
display(customers.columns)

print("\nDataset Numerical Statistics: ")
display(customers.describe())

First 20 rows: 


,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate
0,9207fa75-5758-48d1-94ad-19c041e0520f,51.0,Female,Jensenberg,Low,2022-11-17
1,5fb09cd8-a473-46f7-80bd-6e49cf509078,NaN,Female,Castilloport,High,2020-07-21
2,c139496e-cc89-498a-bd90-1fb4627b6cff,37.0,Male,Lake Jennifertown,NaN,2021-01-01
3,50118139-7264-428f-81cc-a25fddc5d6dd,44.0,Male,Port Carl,Medium,2024-06-10
4,7d1f2bbc-8d16-4fbc-9b37-ece3324e8ed4,50.0,Female,Jessebury,High,2023-08-24
5,2de49c7c-32ae-4ba8-b058-622a090d7094,53.0,Female,Emilyville,Low,2022-02-13
6,89f7de44-e592-43c6-b033-bcbcf24088ba,NaN,Female,South Derekbury,High,2019-12-08
7,9f2128cd-2f2b-4f71-987c-29fd045021f0,NaN,Female,Lake Erin,Medium,2022-04-26
8,fbcd7128-ce04-4799-8c1c-722330507b96,NaN,Female,New Reginabury,NaN,2022-04-17
9,40f5a5dd-a46f-4a5e-8c0f-fe04389ddec8,NaN,Female,Mariaberg,High,2024-02-18



Dataset Information: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3200 entries, 0 to 3199
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   CustomerID   3200 non-null   object
 1   Age          2909 non-null   object
 2   Gender       3200 non-null   object
 3   Location     3200 non-null   object
 4   IncomeLevel  2897 non-null   object
 5   SignupDate   3200 non-null   object
dtypes: object(6)
memory usage: 150.1+ KB


None


Dataset Dimensions: 


(3200, 6)


Dataset Columns: 


Index(['CustomerID', 'Age', 'Gender', 'Location', 'IncomeLevel', 'SignupDate'], dtype='object')


Dataset Numerical Statistics: 


,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate
count,3200,2909,3200,3200,2897,3200
unique,3000,58,2,2696,3,1544
top,1a5115af-2fa7-4198-b838-be2cdeccb768,26.0,Female,Port Michael,High,2024-06-08
freq,2,72,1610,6,1004,9


##### exact duplicate values

In [107]:
# check duplicates
duplicate_customers = customers.duplicated().sum()

print("Number of Exact Duplicate Rows: ")
print(duplicate_customers)

# inspect duplicates
display(customers[customers.duplicated(keep=False)])

Number of Exact Duplicate Rows: 
177


,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate
9,40f5a5dd-a46f-4a5e-8c0f-fe04389ddec8,NaN,Female,Mariaberg,High,2024-02-18
18,8de7560f-370d-4cfd-8135-20a3de237264,31.0,Female,Colemanstad,Low,2020-12-01
23,fd63f236-3ec3-401d-9fa1-53bda0bd7c0e,NaN,Male,Thomasbury,Low,2021-05-29
54,c9f831b4-43cf-4ed9-9b47-09542e239c70,53.0,Female,East Ryanbury,Medium,2023-10-12
56,348c5ab5-992d-4780-ba71-d1a5e3818191,46.0,Female,New Jesseside,Low,2023-09-24
...,...,...,...,...,...,...
3194,40c2909d-983b-4c95-a737-2022acba98ee,28.0,Male,Josephborough,High,2020-02-14
3196,0ef84cc6-0b4c-4c56-bdaa-d6b725ccc08e,40.0,Male,Lisaside,Medium,2022-07-23
3197,0ede97d1-bb46-452d-893e-b63f37da4fa8,NaN,Female,North Christinaport,High,2021-10-16
3198,43ec68f5-4305-4b40-8ced-a187d00bb3f5,26.0,Female,Carrilloport,Low,2022-09-26


In [108]:
# remove duplicates
customer_cleaned = customers.drop_duplicates().copy()

# verify
print("Duplicates Remaining:")
print(customer_cleaned.duplicated().sum())

Duplicates Remaining:
0


#### missing and invalid values

In [109]:
# check missing values
missing_values = customer_cleaned.isnull().sum()
print("Missing Values:")
print(missing_values)

Missing Values:
CustomerID       0
Age            276
Gender           0
Location         0
IncomeLevel    285
SignupDate       0
dtype: int64


In [110]:
# inspect data types and check numerical statistics
print("Data Types:")
print(customer_cleaned.dtypes)

print("\nNumerical Statistics:")
display(customer_cleaned.describe())

Data Types:
CustomerID     object
Age            object
Gender         object
Location       object
IncomeLevel    object
SignupDate     object
dtype: object

Numerical Statistics:


,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate
count,3023,2747,3023,3023,2738,3023
unique,3000,58,2,2696,3,1544
top,08995799-8d0f-40d6-986b-452074709b5f,62.0,Female,Port Michael,High,2024-06-08
freq,2,70,1528,5,946,9


In [111]:
# check categorical inconsistencies
for column in customer_cleaned.select_dtypes(include="object").columns:
    print(f"\n{column}")
    print(customer_cleaned[column].value_counts(dropna=False))


CustomerID
CustomerID
08995799-8d0f-40d6-986b-452074709b5f    2
c2351d11-59fe-4630-a6df-e01bc343de33    2
50465bde-6a9a-466d-804a-a8f046809a23    2
9612ee49-09dd-4f55-9524-9eb341efa2b9    2
b72f2a59-aab0-447d-bb31-9bdc1405cc72    2
                                       ..
05e09646-040f-425a-b5b3-d6f113e0c75b    1
8eb5a52b-e337-426c-8005-a95cb7c02e85    1
09607bcf-49c0-4079-b620-744d5584bc15    1
3f143ceb-9a9f-4305-90d6-fa7a298ce6de    1
db9c2206-4487-47e9-8fb0-0ddc81b7310c    1
Name: count, Length: 3000, dtype: int64

Age
Age
NaN        276
62.0        70
57.0        67
34.0        64
26.0        64
68.0        62
55.0        58
43.0        56
70.0        56
61.0        56
36.0        55
18.0        55
25.0        55
52.0        54
27.0        54
37.0        54
59.0        54
45.0        54
53.0        53
58.0        53
64.0        52
23.0        52
56.0        51
38.0        51
28.0        50
49.0        50
46.0        50
66.0        50
40.0        49
63.0        49
67.0        49
4

##### invalid age values

In [112]:
# convert age from string to integer
customer_cleaned["Age"] = pd.to_numeric(customer_cleaned["Age"], errors="coerce")

# check numerical statistics
print("Numerical Statistics of Age Column:")
display(customer_cleaned["Age"].describe())

# check again number of missing values
print("Number of Missing Values:")
missing_age = customer_cleaned["Age"].isnull().sum().copy()
print(missing_age)

Numerical Statistics of Age Column:


count    2713.000000
mean       45.421673
std        19.731360
min        -1.000000
25%        31.000000
50%        45.000000
75%        58.000000
max       150.000000
Name: Age, dtype: float64

Number of Missing Values:
310


invalid age values are converted to missing prior to filling or dropping them

In [113]:
# count how many invalid age values are in the dataset
invalid_age = customer_cleaned[(customer_cleaned["Age"] < 0) | (customer_cleaned["Age"] > 100)]
print("Invalid Ages:", len(invalid_age))
display(invalid_age)

Invalid Ages: 63


,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate
10,c96a5ee9-f1a6-416a-adc6-1c8b128c7399,150.0,Male,Hansontown,High,2020-09-17
48,b1d6a46b-214c-4592-9688-82d0fd977bb2,150.0,Female,North Cindyhaven,Low,2020-06-28
64,52303d17-9d24-4c20-89e3-7bcfa92140c9,150.0,Male,Spencermouth,NaN,2022-06-25
72,7a461a97-18fd-4686-ab3f-adaded72069d,150.0,Male,Pageside,High,2024-04-16
99,5ba7a03b-0d2f-4fc3-bd36-7a675c1ac61e,150.0,Female,East Ryanchester,High,2021-02-18
...,...,...,...,...,...,...
2916,730f04f4-652c-4eb0-ab28-4e63ca84e166,-1.0,Female,Meganhaven,Low,2024-04-12
2975,ba3f79ad-b3ab-47e5-8803-22cb4ea8cf13,150.0,Female,Rodriguezside,NaN,2022-08-10
3004,3814b8e4-cade-4665-86af-dda3e3628f17,150.0,Female,Port Bonnieshire,Medium,2021-06-25
3132,6b2f8f03-ab9e-48a6-a07e-10e794b04fe0,-1.0,Male,Lake Karen,Low,2023-04-04


ages below 0 or above 100 were treated as invalid because they fall outside a reasonable customer age range for this dataset

In [114]:
# convert invalid values to missing values
customer_cleaned.loc[(customer_cleaned["Age"] < 0) | (customer_cleaned["Age"] > 100), "Age"] = pd.NA

# then validate if the minimum and maximum age values are much more specific
display(customer_cleaned["Age"].describe())
print("\nMissing Ages:", customer_cleaned["Age"].isnull().sum())

count    2650.000000
mean       44.644906
std        15.405637
min        18.000000
25%        31.000000
50%        45.000000
75%        58.000000
max        70.000000
Name: Age, dtype: float64


Missing Ages: 373


##### income level column


In [115]:
# inspect missing values
print(customer_cleaned["IncomeLevel"].value_counts(dropna=False))

IncomeLevel
High      946
Low       923
Medium    869
NaN       285
Name: count, dtype: int64


285 missing values from incomelevel column cannot be replaced by 'high', despite being the mode, since it would artificially increase the high income group which could affect later eda and visualizations

In [116]:
# fill missing values with 'unknown'
customer_cleaned["IncomeLevel"] = customer_cleaned["IncomeLevel"].fillna("Unknown")

# verify
print(customer_cleaned["IncomeLevel"].value_counts(dropna=False))

IncomeLevel
High       946
Low        923
Medium     869
Unknown    285
Name: count, dtype: int64


missing values are replaced with 'unknown'

##### gender column

In [117]:
# inspect
print(customer_cleaned["Gender"].value_counts(dropna=False))

Gender
Female    1528
Male      1495
Name: count, dtype: int64


##### location column

In [118]:
# inspect
print(customer_cleaned["Location"].value_counts(20))

Location
Port Michael     0.001654
West Michael     0.001323
West Robert      0.001323
Michaelton       0.001323
Port Eric        0.001323
                   ...   
Savagestad       0.000331
Samanthaville    0.000331
Donnahaven       0.000331
Sylviabury       0.000331
Stacyshire       0.000331
Name: proportion, Length: 2696, dtype: float64


In [119]:
# clean whitespace
customer_cleaned["Location"] = customer_cleaned["Location"].str.strip()

##### signup date column

In [120]:
# convert to mixed date type since signup date column has different date formats
customer_cleaned["SignupDate"] = pd.to_datetime(customer_cleaned["SignupDate"], format="mixed", dayfirst=True, errors="coerce")

# verify
print(customer_cleaned["SignupDate"].dtype)

datetime64[ns]


In [121]:
print("Missing/invalid SignupDate values:",
      customer_cleaned["SignupDate"].isna().sum())

Missing/invalid SignupDate values: 0


In [122]:
# inspect
customer_cleaned["SignupDate"].isnull().sum()

np.int64(0)

##### customer id column

In [123]:
# count the number of missing values in each row
customer_cleaned["MissingValues"] = customer_cleaned.isnull().sum(axis=1)

# sort
customer_cleaned = customer_cleaned.sort_values(by="MissingValues")

In [124]:
# investigate duplicate customer id
duplicate_customers = customer_cleaned[customer_cleaned.duplicated(subset='CustomerID', keep=False)]

display(duplicate_customers.sort_values("CustomerID"))

,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate,MissingValues
1536,08995799-8d0f-40d6-986b-452074709b5f,61.0,Male,New Lukeville,Medium,2019-11-15,0
3024,08995799-8d0f-40d6-986b-452074709b5f,61.0,Male,New Lukeville,Medium,2020-03-28,0
3085,1e9e7dcb-e612-46d3-bae6-0c2f47fa0220,44.0,Female,Brianmouth,Medium,2019-11-07,0
304,1e9e7dcb-e612-46d3-bae6-0c2f47fa0220,44.0,Female,Brianmouth,Medium,2022-07-28,0
754,29436bcf-d6f8-44d1-a856-0794abb562ab,68.0,Female,New Fred,Low,2021-09-01,0
3083,29436bcf-d6f8-44d1-a856-0794abb562ab,68.0,Female,New Fred,Low,2020-03-12,0
2106,2cd16dfc-d7b0-47ec-9999-b018daea7f8d,NaN,Female,Lake Carloshaven,Medium,2024-05-02,1
3051,2cd16dfc-d7b0-47ec-9999-b018daea7f8d,52.0,Female,Lake Carloshaven,Medium,2024-05-02,0
1238,3814b8e4-cade-4665-86af-dda3e3628f17,32.0,Female,Port Bonnieshire,Medium,2021-06-25,0
3004,3814b8e4-cade-4665-86af-dda3e3628f17,NaN,Female,Port Bonnieshire,Medium,2021-06-25,1


In [125]:
# retain the most complete record for each customer id
customer_cleaned = customer_cleaned.sort_values(["CustomerID", "MissingValues"],ascending=[True, True])

# inspect before deleting, making sure the first record that appears is the best one
display(customer_cleaned[customer_cleaned.duplicated(subset="CustomerID", keep=False)])

,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate,MissingValues
1536,08995799-8d0f-40d6-986b-452074709b5f,61.0,Male,New Lukeville,Medium,2019-11-15,0
3024,08995799-8d0f-40d6-986b-452074709b5f,61.0,Male,New Lukeville,Medium,2020-03-28,0
3085,1e9e7dcb-e612-46d3-bae6-0c2f47fa0220,44.0,Female,Brianmouth,Medium,2019-11-07,0
304,1e9e7dcb-e612-46d3-bae6-0c2f47fa0220,44.0,Female,Brianmouth,Medium,2022-07-28,0
3083,29436bcf-d6f8-44d1-a856-0794abb562ab,68.0,Female,New Fred,Low,2020-03-12,0
754,29436bcf-d6f8-44d1-a856-0794abb562ab,68.0,Female,New Fred,Low,2021-09-01,0
3051,2cd16dfc-d7b0-47ec-9999-b018daea7f8d,52.0,Female,Lake Carloshaven,Medium,2024-05-02,0
2106,2cd16dfc-d7b0-47ec-9999-b018daea7f8d,NaN,Female,Lake Carloshaven,Medium,2024-05-02,1
1238,3814b8e4-cade-4665-86af-dda3e3628f17,32.0,Female,Port Bonnieshire,Medium,2021-06-25,0
3004,3814b8e4-cade-4665-86af-dda3e3628f17,NaN,Female,Port Bonnieshire,Medium,2021-06-25,1


In [126]:
# retain the first appearing record
customer_cleaned = customer_cleaned.drop_duplicates(subset="CustomerID", keep="first")

# remove temporary column
customer_cleaned = customer_cleaned.drop(columns=["MissingValues"])

# verify
print("Duplicate CustomerIDs:", customer_cleaned["CustomerID"].duplicated().sum())

Duplicate CustomerIDs: 0


##### missing age values

In [128]:
# check how many records have missing age values after deduplication
print("Remaining missing Age values:")
print(customer_cleaned["Age"].isna().sum())

Remaining missing Age values:
362


since a substantial number of records still have missing age values after deduplication, these missing values will be filled with the median value

In [130]:
# inspect numerical statistics of age column
customer_cleaned["Age"].describe()

count    2638.000000
mean       44.618271
std        15.412377
min        18.000000
25%        31.000000
50%        45.000000
75%        58.000000
max        70.000000
Name: Age, dtype: float64

In [134]:
# fill missing age values with median
customer_cleaned["Age"] = customer_cleaned["Age"].fillna(customer_cleaned["Age"].median())

# verify
print("Missing Age Values:", customer_cleaned["Age"].isna().sum())
display(customer_cleaned["Age"].describe())

Missing Age Values: 0


count    3000.000000
mean       44.664333
std        14.452818
min        18.000000
25%        33.750000
50%        45.000000
75%        57.000000
max        70.000000
Name: Age, dtype: float64

### customer transactions dataset preprocessing